File: backend/routers/analysis.py
from datetime import UTC, datetime, timedelta

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from database import get_session
from db_models import AnalysisResultRecord, ExperimentRecord
from engines.ai_engine import AiEngine
from models.experiment import ExperimentConfig, ExperimentResult

router = APIRouter()
ai_engine = AiEngine()


def _record_to_result(rec: ExperimentRecord) -> ExperimentResult:
    """Convert a DB record to an ExperimentResult Pydantic model."""
    from models.experiment import ExperimentPhase, ExperimentStatus

    return ExperimentResult(
        experiment_id=rec.id,
        config=ExperimentConfig(**rec.config),
        status=ExperimentStatus(rec.status),
        phase=ExperimentPhase(rec.phase),
        started_at=rec.started_at,
        completed_at=rec.completed_at,
        steady_state=rec.steady_state,
        hypothesis=rec.hypothesis,
        injection_result=rec.injection_result,
        observations=rec.observations,
        rollback_result=rec.rollback_result,
        error=rec.error,
        ai_insights=rec.ai_insights,
    )


@router.post("/experiment/{experiment_id}")
async def analyze_experiment(
    experiment_id: str,
    session: AsyncSession = Depends(get_session),
):
    """Analyze a completed experiment using AI."""
    rec = await session.get(ExperimentRecord, experiment_id)
    if not rec:
        raise HTTPException(status_code=404, detail="Experiment not found")

    exp = _record_to_result(rec)
    result = await ai_engine.analyze_experiment(
        experiment_data=exp.model_dump(),
        steady_state=exp.steady_state or {},
        observations=exp.observations or {},
    )

    # Persist analysis result
    analysis_rec = AnalysisResultRecord(
        experiment_id=experiment_id,
        severity=result.severity,
        root_cause=result.root_cause,
        confidence=result.confidence,
        recommendations=result.recommendations,
        resilience_score=result.resilience_score,
    )
    session.add(analysis_rec)
    await session.commit()

    return result.model_dump()


@router.post("/hypotheses")
async def generate_hypotheses(request: dict):
    """Generate failure hypotheses for a target."""
    topology = request.get("topology", {})
    target = request.get("target", "")
    chaos_type = request.get("chaos_type", "")

    hypothesis = await ai_engine.generate_hypothesis(topology, target, chaos_type)
    return {"hypothesis": hypothesis}


@router.post("/resilience-score")
async def calculate_resilience_score(request: dict):
    """Calculate resilience score from experiment history."""
    experiments_data = request.get("experiments", [])
    score = await ai_engine.calculate_resilience_score(experiments_data)
    return score


@router.post("/report")
async def generate_report(request: dict):
    """Generate a human-readable experiment report."""
    experiment_data = request.get("experiment", {})
    analysis = request.get("analysis")
    report = await ai_engine.generate_report(experiment_data, analysis)
    return {"report": report}


@router.post("/generate-experiments")
async def generate_experiments(request: dict):
    """Generate experiment configs from topology using AI."""
    topology = request.get("topology", {})
    target_namespace = request.get("target_namespace", "default")
    count = request.get("count", 3)

    experiments = await ai_engine.generate_experiments(topology, target_namespace, count)
    return {"experiments": experiments, "count": len(experiments)}


@router.post("/nl-experiment")
async def nl_experiment(request: dict):
    """Convert natural language to ExperimentConfig."""
    text = request.get("text", "").strip()
    if not text:
        raise HTTPException(status_code=400, detail="Text is required")

    topology = request.get("topology")
    config = await ai_engine.parse_natural_language(text, topology)
    return config


@router.get("/resilience-trend")
async def resilience_trend(
    namespace: str | None = Query(default=None),
    days: int = Query(default=30, ge=1, le=365),
    session: AsyncSession = Depends(get_session),
):
    """Get resilience score trend from analysis history."""
    since = datetime.now(UTC) - timedelta(days=days)

    query = select(AnalysisResultRecord).where(AnalysisResultRecord.created_at >= since)

    if namespace:
        # Join with experiments to filter by namespace in config
        query = (
            select(AnalysisResultRecord)
            .join(ExperimentRecord, AnalysisResultRecord.experiment_id == ExperimentRecord.id)
            .where(
                AnalysisResultRecord.created_at >= since,
                ExperimentRecord.config["target_namespace"].as_string() == namespace,
            )
        )

    query = query.order_by(AnalysisResultRecord.created_at.asc())
    result = await session.execute(query)
    records = result.scalars().all()

    return {
        "trend": [
            {
                "experiment_id": r.experiment_id,
                "resilience_score": r.resilience_score,
                "severity": r.severity,
                "created_at": r.created_at.isoformat() if r.created_at else None,
            }
            for r in records
        ],
        "count": len(records),
        "period_days": days,
        "namespace": namespace,
    }


@router.get("/resilience-trend/summary")
async def resilience_trend_summary(
    namespace: str | None = Query(default=None),
    days: int = Query(default=30, ge=1, le=365),
    session: AsyncSession = Depends(get_session),
):
    """Get AI-generated summary of resilience score trend."""
    since = datetime.now(UTC) - timedelta(days=days)

    query = (
        select(AnalysisResultRecord)
        .where(AnalysisResultRecord.created_at >= since)
        .order_by(AnalysisResultRecord.created_at.asc())
    )
    result = await session.execute(query)
    records = result.scalars().all()

    experiments_data = [
        {
            "experiment_id": r.experiment_id,
            "resilience_score": r.resilience_score,
            "severity": r.severity,
            "root_cause": r.root_cause,
        }
        for r in records
    ]

    summary = await ai_engine.calculate_resilience_score(experiments_data)
    return {
        "summary": summary,
        "data_points": len(records),
        "period_days": days,
    }

File: backend-go/internal/handler/analysis.go
package handler

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strconv"
	"time"

	"github.com/chaosduck/backend-go/internal/db"
	"github.com/gin-gonic/gin"
	"github.com/jackc/pgx/v5/pgtype"
)

// AnalysisHandler proxies AI analysis requests to the Python AI microservice
type AnalysisHandler struct {
	queries      *db.Queries
	aiServiceURL string
	httpClient   *http.Client
}

// NewAnalysisHandler creates a new AnalysisHandler
func NewAnalysisHandler(queries *db.Queries, aiServiceURL string) *AnalysisHandler {
	return &AnalysisHandler{
		queries:      queries,
		aiServiceURL: aiServiceURL,
		httpClient:   &http.Client{Timeout: 60 * time.Second},
	}
}

// AnalyzeExperiment proxies to AI service for experiment analysis
func (h *AnalysisHandler) AnalyzeExperiment(c *gin.Context) {
	experimentID := c.Param("experiment_id")

	rec, err := h.queries.GetExperiment(c.Request.Context(), experimentID)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"detail": "Experiment not found"})
		return
	}

	result := recordToResult(rec)

	body := map[string]any{
		"experiment_data": result,
		"steady_state":    result.SteadyState,
		"observations":    result.Observations,
	}

	resp, err := h.proxyToAI("/analyze", body)
	if err != nil {
		c.JSON(http.StatusBadGateway, gin.H{"detail": fmt.Sprintf("AI service error: %v", err)})
		return
	}

	// Persist analysis result if we got one
	if severity, ok := resp["severity"].(string); ok {
		rootCause, _ := resp["root_cause"].(string)
		confidence, _ := resp["confidence"].(float64)
		resilienceScore, _ := resp["resilience_score"].(float64)
		recsJSON, _ := json.Marshal(resp["recommendations"])

		h.queries.CreateAnalysisResult(c.Request.Context(), db.CreateAnalysisResultParams{
			ExperimentID:    experimentID,
			Severity:        severity,
			RootCause:       rootCause,
			Confidence:      confidence,
			Recommendations: recsJSON,
			ResilienceScore: pgtype.Float8{Float64: resilienceScore, Valid: true},
		})
	}

	c.JSON(http.StatusOK, resp)
}

// GenerateHypotheses proxies to AI service
func (h *AnalysisHandler) GenerateHypotheses(c *gin.Context) {
	var body map[string]any
	if err := c.ShouldBindJSON(&body); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"detail": err.Error()})
		return
	}

	resp, err := h.proxyToAI("/hypotheses", body)
	if err != nil {
		c.JSON(http.StatusBadGateway, gin.H{"detail": fmt.Sprintf("AI service error: %v", err)})
		return
	}
	c.JSON(http.StatusOK, resp)
}

// CalculateResilienceScore proxies to AI service
func (h *AnalysisHandler) CalculateResilienceScore(c *gin.Context) {
	var body map[string]any
	if err := c.ShouldBindJSON(&body); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"detail": err.Error()})
		return
	}

	resp, err := h.proxyToAI("/resilience-score", body)
	if err != nil {
		c.JSON(http.StatusBadGateway, gin.H{"detail": fmt.Sprintf("AI service error: %v", err)})
		return
	}
	c.JSON(http.StatusOK, resp)
}

// GenerateReport proxies to AI service
func (h *AnalysisHandler) GenerateReport(c *gin.Context) {
	var body map[string]any
	if err := c.ShouldBindJSON(&body); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"detail": err.Error()})
		return
	}

	resp, err := h.proxyToAI("/report", body)
	if err != nil {
		c.JSON(http.StatusBadGateway, gin.H{"detail": fmt.Sprintf("AI service error: %v", err)})
		return
	}
	c.JSON(http.StatusOK, resp)
}

// GenerateExperiments proxies to AI service
func (h *AnalysisHandler) GenerateExperiments(c *gin.Context) {
	var body map[string]any
	if err := c.ShouldBindJSON(&body); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"detail": err.Error()})
		return
	}

	resp, err := h.proxyToAI("/generate-experiments", body)
	if err != nil {
		c.JSON(http.StatusBadGateway, gin.H{"detail": fmt.Sprintf("AI service error: %v", err)})
		return
	}
	c.JSON(http.StatusOK, resp)
}

// NLExperiment proxies natural language experiment creation to AI service
func (h *AnalysisHandler) NLExperiment(c *gin.Context) {
	var body map[string]any
	if err := c.ShouldBindJSON(&body); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"detail": err.Error()})
		return
	}

	text, _ := body["text"].(string)
	if text == "" {
		c.JSON(http.StatusBadRequest, gin.H{"detail": "Text is required"})
		return
	}

	resp, err := h.proxyToAI("/nl-experiment", body)
	if err != nil {
		c.JSON(http.StatusBadGateway, gin.H{"detail": fmt.Sprintf("AI service error: %v", err)})
		return
	}
	c.JSON(http.StatusOK, resp)
}

// ResilienceTrend returns resilience score trend from DB
func (h *AnalysisHandler) ResilienceTrend(c *gin.Context) {
	namespace := c.Query("namespace")
	daysStr := c.DefaultQuery("days", "30")
	days, err := strconv.Atoi(daysStr)
	if err != nil || days < 1 || days > 365 {
		days = 30
	}

	since := time.Now().UTC().Add(-time.Duration(days) * 24 * time.Hour)
	sinceTS := pgtype.Timestamptz{Time: since, Valid: true}

	var records []db.AnalysisResult
	if namespace != "" {
		records, err = h.queries.ListAnalysisResultsSinceByNamespace(c.Request.Context(), db.ListAnalysisResultsSinceByNamespaceParams{
			Since:     sinceTS,
			Namespace: namespace,
		})
	} else {
		records, err = h.queries.ListAnalysisResultsSince(c.Request.Context(), sinceTS)
	}
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"detail": err.Error()})
		return
	}

	trend := make([]map[string]any, 0, len(records))
	for _, r := range records {
		entry := map[string]any{
			"experiment_id":   r.ExperimentID,
			"severity":        r.Severity,
		}
		if r.ResilienceScore.Valid {
			entry["resilience_score"] = r.ResilienceScore.Float64
		}
		if r.CreatedAt.Valid {
			entry["created_at"] = r.CreatedAt.Time.Format(time.RFC3339)
		}
		trend = append(trend, entry)
	}

	c.JSON(http.StatusOK, gin.H{
		"trend":       trend,
		"count":       len(trend),
		"period_days": days,
		"namespace":   namespace,
	})
}

// ResilienceTrendSummary returns AI-generated summary of the trend
func (h *AnalysisHandler) ResilienceTrendSummary(c *gin.Context) {
	daysStr := c.DefaultQuery("days", "30")
	days, err := strconv.Atoi(daysStr)
	if err != nil || days < 1 || days > 365 {
		days = 30
	}

	since := time.Now().UTC().Add(-time.Duration(days) * 24 * time.Hour)
	sinceTS := pgtype.Timestamptz{Time: since, Valid: true}

	records, err := h.queries.ListAnalysisResultsSince(c.Request.Context(), sinceTS)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"detail": err.Error()})
		return
	}

	experimentsData := make([]map[string]any, 0, len(records))
	for _, r := range records {
		entry := map[string]any{
			"experiment_id": r.ExperimentID,
			"severity":      r.Severity,
			"root_cause":    r.RootCause,
		}
		if r.ResilienceScore.Valid {
			entry["resilience_score"] = r.ResilienceScore.Float64
		}
		experimentsData = append(experimentsData, entry)
	}

	body := map[string]any{"experiments": experimentsData}
	resp, err := h.proxyToAI("/resilience-score", body)
	if err != nil {
		c.JSON(http.StatusBadGateway, gin.H{"detail": fmt.Sprintf("AI service error: %v", err)})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"summary":     resp,
		"data_points": len(records),
		"period_days": days,
	})
}

// proxyToAI sends a JSON POST request to the AI microservice
func (h *AnalysisHandler) proxyToAI(path string, body any) (map[string]any, error) {
	jsonBody, err := json.Marshal(body)
	if err != nil {
		return nil, fmt.Errorf("marshal body: %w", err)
	}

	resp, err := h.httpClient.Post(
		h.aiServiceURL+path,
		"application/json",
		bytes.NewReader(jsonBody),
	)
	if err != nil {
		return nil, fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("read response: %w", err)
	}

	if resp.StatusCode >= 400 {
		return nil, fmt.Errorf("AI service returned %d: %s", resp.StatusCode, string(respBody))
	}

	var result map[string]any
	if err := json.Unmarshal(respBody, &result); err != nil {
		return nil, fmt.Errorf("parse response: %w", err)
	}

	return result, nil
}
